homeassistant/components/tasmota/discovery.py

```diff
@@ -149,22 +123,29 @@ async def _subscribe_topics(sub_state, topics):
                     async def _unsubscribe_topics(sub_state):
                         return await async_unsubscribe_topics(hass, sub_state)
 
-                    entity = get_entity(entity_config, component_key)
-                    entity.set_mqtt_callbacks(
+                    tasmota_entity = tasmota_get_entity(
+                        tasmota_entity_config, component_key
+                    )
+                    tasmota_entity.set_mqtt_callbacks(
                         _publish, _subscribe_topics, _unsubscribe_topics
                     )
                     async_dispatcher_send(
                         hass,
                         TASMOTA_DISCOVERY_ENTITY_NEW.format(component),
-                        entity,
+                        tasmota_entity,
                         discovery_hash,
                     )
 
     hass.data[DATA_CONFIG_ENTRY_LOCK] = asyncio.Lock()
     hass.data[CONFIG_ENTRY_IS_SETUP] = set()
 
-    hass.data[DISCOVERY_UNSUBSCRIBE] = await mqtt.async_subscribe(
-        hass, f"{discovery_topic}/#", async_discovery_message_received, 0
+    async def _subscribe(topic, msg_callback, *args):
+        hass.data[DISCOVERY_UNSUBSCRIBE] = await mqtt.async_subscribe(
```

Comment:
I'd expect the library to clean up too.

Reply:
Consider making a tasmota mqtt client class in the library that accepts the publish and subscribe callbacks from the mqtt transport provider (here the mqtt integration). Then use the client instance in the other tasmota discovery functions. Either pass the client as an argument to each function or put the functions as methods on the client.

Then we don't need to keep passing the transport callbacks to each tasmota discovery function.

This will make it easier to test the library too, since you then just need to mock the client once.